common/lib/xmodule/xmodule/modulestore/django.py

```diff
@@ -243,10 +245,47 @@ class ModuleI18nService(object):
     i18n service.
 
     """
+    def __init__(self, block):
+        """
+        Store a reference to the block currently being serviced by this runtime.  We'll use this
+        reference later on for things like locating the block's translation domain (PO+MO files)
+        """
+        self.block = block
 
     def __getattr__(self, name):
         return getattr(django.utils.translation, name)
 
+    def ugettext(self, string):
+        """
+        This operation is a proxy for django.utils.translation.ugettext, which is itself a proxy for
+        the GNU gettext ugettext operation.  Here we attempts to look up the provided string in the
+        XBlock's own domain translation catalog, currently expected to be:
+            <xblock_root>/conf/locale/<language>/LC_MESSAGES/<domain>.po
+        If ModuleI18nService can't locate the domain translation catalog then we fall-back onto
+        django.utils.translation.ugettext, which will attempt to find a matching string in the
+        LMS' own domain translation catalog -- effectively achieving translation by coincidence.
+        """
+        # The translation workflow should only execute if there's an actual string to look up
+        # If gettext processes an empty string the PO file header information will oddly be returned
+        translated_string = unicode(string)
+        if translated_string and self.block:
+            try:
+                xblock_resource = self.block.__class__.unmixed_class.__module__
+                xblock_locale_dir = '/conf/locale'
+                xblock_locale_path = resource_filename(xblock_resource, xblock_locale_dir)
```

Comment:
We'll need to write instructions for XBlocks, and the layout will look different than the way edx-platform and edx-ora2 both do it.  The "conf/locale" directory won't be at the root of the project, but inside one of the package directories.